bundler/lib/bundler/gem_version_promoter.rb

```diff
@@ -78,8 +85,8 @@ def filter_dep_specs(specs, package)
 
           must_match = minor? ? [0] : [0, 1]
 
-          matches = must_match.map {|idx| gsv.segments[idx] == lsv.segments[idx] }
-          matches.uniq == [true] ? (gsv >= lsv) : false
+          all_match = must_match.all? {|idx| gsv.segments[idx] == lsv.segments[idx] }
+          all_match && gsv >= lsv
```

Comment:
Nice! Much easier to understand. Normally I extract these little refactorings/improvements to their own commit, but introduce them with the same PR where I come up with them. But I'm happy if you move them to a separate PR too!

By the way, another improvement here would be to return early if `locked_version & !major?` is not met, rather than checking that every time in the `select`.